File: DP/Coin Change/dp.py
# https://leetcode.com/problems/coin-change/
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if (amount==0): return 0
        # coins.sort()
        
        dp=[]
        dp.append(0)
        for i in range(1,amount+1):
            val=float('inf')
            for x in coins:
                if (i-x<0): pass
                elif dp[i-x]+1<val: val=dp[i-x]+1
            dp.append(val)
        
        # print(dp)
        if dp[-1]==float('inf'): return -1
        else: return dp[-1]
        

File: DP/Coin Change/dp.cpp
# https://leetcode.com/problems/coin-change/
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount==0) return 0;
        vector <int> dp(amount+1,INT_MAX);
        dp[0]=0;
        for (int i=0; i <= amount;i++){
            for (auto v: coins){
                if (i-v>=0 && dp[i-v]!=INT_MAX && dp[i]> dp[i-v]+1){
                    dp[i]=dp[i-v]+1;
                }
            }
        }
        return ((dp[amount] == INT_MAX) ? -1 : dp[amount]);
    }
};
        